Foliage osm data fixing

fyodorovvictor opened the pull request:
Fixing unclosed relation contours and converting road lines into contours to remove them from foliage mask

fyodorovvictor committed:
Removing building and road contours from foliage mask

fyodorovvictor committed:
Merge branch 'MapDataLoaderFoliage' into FoliageOsmDataFixing

# Conflicts:
#	GeoTempCore/Source/GeoTempLoaders/Private/OSM/LoaderFoliageOsm.cpp

Added relation contours fixing

fyodorovvictor committed:
Merge branch 'develop' into FoliageOsmDataFixing

fyodorovvictor committed:
Fixing foliage suppression by roads

fyodorovvictor committed:
Merge branch 'develop' into FoliageOsmDataFixing

Xanadar89 commented on GeoTempCore/Source/GeoTempCore/Private/Contour.cpp:
```diff
@@ -133,6 +133,12 @@ bool FContour::IsNotClockwise(bool inReverse)
 
+bool FContour::IsClosed()
+{
+	return (Points.Last() - Points[0]).Size2D() <= 1;
```
а не слишком ли большой эпсилон, аж 1?

fyodorovvictor replied:
Нужно уточнить у Кирилла, так как он писал эту логику. Я ее лишь обернул.

Xanadar89 commented on GeoTempCore/Source/GeoTempLoaders/Private/OSM/LoaderFoliageOsm.cpp:
```diff
@@ -1,5 +1,171 @@
+#include "LoaderHelper.h"
+
+
+inline const double signedArea(const TArray<FVector> &p)
```
Почему имя функции с маленькой буквы?

fyodorovvictor replied:
Убрал функцию за ненадобностью

Xanadar89 commented on GeoTempCore/Source/GeoTempLoaders/Private/OSM/LoaderFoliageOsm.cpp:
```diff
@@ -1,5 +1,171 @@
+}
+
+
+inline const TArray<FVector> border(const TArray<FVector> &p, double thickness)
```
и тут?
и inline убирай везде в cpp файлах

fyodorovvictor replied:
Также убрал функцию за ненадобностью

fyodorovvictor committed:
Implementing requested changes
Minor refactoring
Adding comments to GeoTempVis plugin (by mistake in current branch and not in separate one)

fyodorovvictor committed:
Moving contour fixing to LoaderHelper
Adding contour fixing to LoaderBuildingsOsm

fyodorovvictor committed:
Merge branch 'develop' into FoliageOsmDataFixing

fyodorovvictor committed:
Merging with develop branch

Xanadar89 commented on GeoTempCore/Source/GeoTempLoaders/Private/LoaderHelper.cpp:
```diff
@@ -56,3 +56,91 @@ FRoadNetwork ULoaderHelper::ConstructRoadNetwork(TArray<FRoadSegment> inRoadSegm
+			closedContours.Add(inUnclosedContours[0]);
+			inUnclosedContours.RemoveAt(0);
+			continue;
+		}
```
если это обязательно надо выполнить на первом проходе, то что мешает вынести это за цикл, выше
if(inUnclosedContours.Num() > 0) 
{
    closedContours.Add(inUnclosedContours[0]);
	inUnclosedContours.RemoveAt(0);
}
while (inUnclosedContours.Num() > 0)
{
    ...
}
а так у тебя лишняя проверка которая выполняется на каждом цикле

Xanadar89 commented on GeoTempCore/Source/GeoTempLoaders/Private/OSM/LoaderBuildingsOsm.cpp:
```diff
@@ -256,13 +263,27 @@ TArray<FBuilding> ULoaderBuildingsOsm::GetBuildings_Implementation()
+						{
+							UnclosedInnerContours.Add(contour);
+						}
 					}
```
у element.Value могут быть другие значения кроме inner outer?
Потому что если их только эти два, то этот код можно очень хорошо переписать

Xanadar89 commented on GeoTempCore/Source/GeoTempLoaders/Private/OSM/LoaderBuildingsOsm.cpp:
```diff
@@ -333,17 +360,33 @@ TArray<FBuilding> ULoaderBuildingsOsm::GetBuildings_Implementation()
+						{
+							UnclosedInnerContours.Add(contour);
+						}
 					}
```
и тут тоже

Xanadar89 commented on GeoTempCore/Source/GeoTempLoaders/Private/OSM/LoaderFoliageOsm.cpp:
```diff
@@ -40,11 +49,87 @@ TArray<FContourData> ULoaderFoliageOsm::GetFolliage_Implementation()
+
+				FVector radiusDeltas[capDensity + 1];
+
+				for (int j = 1; j < capDensity; j++)
+				{
+					float angle = PI / capDensity * j;
+
+					float x = FMath::Cos(angle);
+					float y = FMath::Sin(angle);
+
+					roadCont.Points.Add(endPoint - FVector(0, 0, 2) + (pointDelta * x + yDelta * y));
+				}
+
+				roadCont.Points.Append({
+					point3,
+					point1
+				});
+
+				for (int j = 1; j < capDensity; j++)
+				{
+					float angle = PI / capDensity * j;
+
+					float x = FMath::Cos(angle);
+					float y = FMath::Sin(angle);
+
+					roadCont.Points.Add(startPoint - FVector(0, 0, 2) - (pointDelta * x + yDelta * y));
+				}
+				
+				roadCont.Points.Append({
+					point0
+				});
+				
+				roadPolygon.Outer.Add(roadCont);
+				roadPolygon.Tags = way->Tags;
+				roadPolygon.Tags.Add(TPair<FString, FString>("Type", "Exclude"));
+
+				polygons.Add(roadPolygon);
+			}
+		}
 	}
```
это что копия кода Влада? так делать нельзя, просто вызови уже существующую функцию

fyodorovvictor replied:
У Влада этот код является частью функции создания меша дороги и не вынесен в отдельную функцию. Даже если я попробую вызывать функцию строительсва меша, то итоговые точки будут не в том порядке, который требуется для создания полигонов.

Xanadar89 commented on GeoTempCore/Source/GeoTempLoaders/Private/OSM/LoaderFoliageOsm.cpp:
```diff
@@ -79,20 +170,42 @@ TArray<FContourData> ULoaderFoliageOsm::GetFolliage_Implementation()
 
 				if (element.Value == "outer")
```
а это копия кода Кирилла?

fyodorovvictor replied:
Я брал за основу код Кирилла, так как в нем была логика чтения полигонов. Если я правильно понимаю, то код отличается тегами и способом их обработки, но сборка полигонов из точек одна и та же.

Xanadar89 commented on GeoTempVis/Source/GeoTempFoliage/Public/CustomFoliageInstancer.h:
```diff
@@ -13,13 +13,30 @@
-	MonoLayered		UMETA(DisplayName = "MonoLayered"),
-	PolyLayered		UMETA(DisplayName = "PolyLayered"),
+	MonoLayered		UMETA(DisplayName = "MonoLayered"), /**
+														* < Place meshes from multiple FoliageMeshes as one layer. 
```
зачем на 25 строчке начинать писать текст, если можно на 24
и далее также

fyodorovvictor committed:
Implementing requested changes
RoadBuilder refactoring: transferring some logic into functions for future use from outside

fyodorovvictor committed:
Merge branch 'develop' into FoliageOsmDataFixing

# Conflicts:
#	GeoTempCore/Source/GeoTempCore/Public/Contour.h
#	GeoTempCore/Source/GeoTempLoaders/Private/OSM/LoaderFoliageOsm.cpp
#	GeoTempCore/Source/GeoTempLoaders/Public/OSM/LoaderFoliageOsm.h
#	GeoTempVis/Content/OSMCase.uasset

Also fixing FoliageLoader/UsageExampleFoliage_OSM.uasset
